File: PY-SDK/public/src/aver_client.py
from anchorpy import Provider, Wallet, Program
from solana.rpc.commitment import Finalized, Confirmed
from solana.rpc import types
from solana.rpc.async_api import AsyncClient
from solana.publickey import PublicKey
from solana.transaction import Transaction
from solana.rpc.api import Client
# from constants import AVER_API_URL_DEVNET, DEVNET_SOLANA_URL, DEFAULT_QUOTE_TOKEN_DEVNET, AVER_PROGRAM_ID_DEVNET_2
from .constants import get_aver_api_endpoint, get_quote_token, get_solana_endpoint, AVER_PROGRAM_ID
from solana.keypair import Keypair
from requests import get, post
from spl.token.instructions import get_associated_token_address, create_associated_token_account
from solana.rpc.commitment import Commitment, Finalized
from .enums import SolanaNetwork

class AverClient():
    """
    Aver Client Class

    Use AverClient to interact with the Aver Program
    """

    connection: AsyncClient
    """AsyncClient"""
    program: Program
    """Program"""
    provider: Provider
    """Provider"""
    aver_api_endpoint: str
    """Endpoint is used to make requests to Aver"""
    solana_network: SolanaNetwork
    """Devnet or Mainnet"""
    quote_token: PublicKey
    """"""
    solana_client: Client
    """"""
    owner: Keypair
    """Pays for transactions onchain"""

    def __init__(
        self, 
        program: Program,
        aver_api_endpoint: str,
        solana_network: SolanaNetwork,
        ):
        """
        Initialises AverClient object. Do not use this function; use AverClient.load() instead

        Args:
            program (Program): Aver program
            aver_api_endpoint (str): Endpoint for Aver API (to be removed soon)
            solana_network (SolanaNetwork): Solana network
        """
        self.connection = program.provider.connection
        self.program = program
        self.provider = program.provider
        self.aver_api_endpoint = aver_api_endpoint
        self.solana_network = solana_network
        self.quote_token = get_quote_token(solana_network)
        self.solana_client = Client(get_solana_endpoint(solana_network))
        self.owner = program.provider.wallet.payer

    @staticmethod
    async def load(
        connection: AsyncClient,
        owner: Keypair or Wallet, 
        opts: types.TxOpts,
        network: SolanaNetwork,
        program_id: PublicKey = AVER_PROGRAM_ID,
        ):
            """
            Initialises an AverClient object

            Args:
                connection (AsyncClient): Solana AsyncClient object
                owner (KeypairorWallet): Keypair to pay transaction costs (and rent costs)
                opts (types.TxOpts): Default options for sending transactions
                network (SolanaNetwork): Solana network
                program_id (PublicKey, optional): Program public key. Defaults to AVER_PROGRAM_ID.

            Raises:
                Exception: Invalid owner argument

            Returns:
                AverClient: Aver Client
            """
            if(isinstance(owner, Wallet)):
                wallet = owner
            elif(isinstance(owner, Keypair)):
                wallet = Wallet(owner)
            else:
                raise Exception('Invalid Owner Passed. Must be Wallet or Keypair')

            provider = Provider(
                connection,
                wallet,
                opts
            )
            api_endpoint = get_aver_api_endpoint(network)
            program = await AverClient.load_program(provider, program_id)

            aver_client = AverClient(program, api_endpoint, network)    
            return aver_client

    @staticmethod
    async def load_program(provider: Provider, program_id: PublicKey = AVER_PROGRAM_ID):
        """
        Loads Aver Program

        Args:
            provider (Provider): Provider
            program_id (PublicKey, optional): Program public key. Defaults to AVER_PROGRAM_ID.

        Raises:
            Exception: Program IDL not loaded

        Returns:
            Program: Program
        """
        idl = await Program.fetch_idl(program_id,provider)
        if(idl):
            program = Program(idl, program_id, provider)
            return program
        else:
            raise Exception('Program IDL not loaded')


    async def close(self):
        """
        Closes connection.

        You should call this in your program's clean-up function(s)
        """
        await self.provider.close()

    #TODO - Move to admin
    async def check_health(self):
        url = self.aver_api_endpoint + '/health?format=json'
        aver_health_response = get(url)

        solana_health_response = await self.provider.connection.get_version()

        return {'api': aver_health_response.json(), 'solana': solana_health_response}


    async def get_or_create_associated_token_account(
            self,
            owner: PublicKey,
            payer: Keypair,
            token_mint: PublicKey = None,
        ):
        """
        Attempts to load an ATA and creates one if not found.

        Args:
            owner (PublicKey): Owner of ATA
            payer (Keypair): Payer of transaction fee and ATA rent (this is recoverable)
            token_mint (PublicKey, optional): ATA token mint public key. Defaults to USDC token according to chosen solana network.

        Returns:
            PublicKey: ATA PublicKey
        """
        token_mint = token_mint if token_mint is not None else self.quote_token
        associated_token_account = get_associated_token_address(owner=owner, mint=token_mint)
        response = await self.provider.connection.get_token_account_balance(associated_token_account)

        if not 'result' in response.keys(): 
            tx = Transaction()
            tx.add(
                create_associated_token_account(
                    payer=payer.public_key,
                    owner=owner,
                    mint=token_mint,
                ),
            )
            signers = [payer]
            response = await self.provider.connection.send_transaction(tx, *signers, opts=self.provider.opts)
            await self.provider.connection.confirm_transaction(response['result'])
            return associated_token_account
        else:
            return associated_token_account

    async def request_lamport_airdrop(self, amount: int, owner: PublicKey):
        """
        Request lamport airdrop. Only works on devnet

        Args:
            amount (int): Lamports requested
            owner (PublicKey): Public key of account to be airdropped

        Returns:
            RPCResponse: Response
        """
        return await self.provider.connection.request_airdrop(
            pubkey=owner, 
            lamports=amount, 
            commitment=self.provider.opts
            )
    

    async def request_ata_balance(self, ata: PublicKey):
        """
        Fetches ATA Balance

        Args:
            ata (PublicKey): ATA public key

        Raises:
            Exception: Error from response

        Returns:
            int: Token balance
        """
        response = await self.provider.connection.get_token_account_balance(ata, self.provider.opts.preflight_commitment)
        if 'error' in response:
            raise Exception(response['error'])
        return int(response['result']['value']['amount'])

    async def request_token_balance(self, mint: PublicKey, owner: PublicKey):
        """"""
        ata = get_associated_token_address(owner, mint)
        return await self.request_ata_balance(ata)

    async def request_lamport_balance(self, owner: PublicKey):
        """
        Fetches Lamport balance

        Args:
            owner (PublicKey): Owner public key

        Raises:
            Exception: Error from response

        Returns:
            int: Lamport balance
        """
        response = await self.provider.connection.get_balance(owner, self.provider.opts.preflight_commitment)
        if 'error' in response:
            raise Exception(response['error'])
        return response['result']['value']

File: TS-SDK/public/src/aver-client.ts
import axios from "axios"
import {
  ConfirmOptions,
  Connection,
  Keypair,
  LAMPORTS_PER_SOL,
  PublicKey,
  Signer,
} from "@solana/web3.js"
import { Program, Provider, Wallet } from "@project-serum/anchor"
import NodeWallet from "@project-serum/anchor/dist/cjs/nodewallet"
import {
  getOrCreateAssociatedTokenAccount,
  createAssociatedTokenAccountInstruction,
  getAssociatedTokenAddress,
  TOKEN_PROGRAM_ID,
  ASSOCIATED_TOKEN_PROGRAM_ID,
} from "@solana/spl-token"
import { AVER_PROGRAM_IDS, getQuoteToken } from "./ids"
import { SolanaNetwork } from "./types"
import { checkIdlHasSameInstructionsAsSdk } from "./checks"

export class AverClient {
  /**
   * Aver Client Class
   *
   * Use AverClient to interact with the Aver Program, Solana network and Aver API
   */

  /**
   * Solana Connection Client
   */
  private _connection: Connection

  /**
   * List of AnchorPy Programs
   */
  private _programs: Program[]

  /**
   * Provider
   */
  private _provider: Provider

  /**
   * Devnet or Mainnet - must correspond to the network provided in the connection object
   */
  private _solanaNetwork: SolanaNetwork

  /**
   * The token mint of the default quote token for markets on this network (i.e. USDC)
   */
  private _quoteTokenMint: PublicKey

   /**
   * The default payer for transactions on-chain, unless one is specified
   */
  private _keypair: Keypair

  /**
   * 
   */
  private _owner: PublicKey

  /**
   * Initialises AverClient object. Do not use this function; use AverClient.load() instead
   *
   * @param {Program} programs - List of Aver program AnchorPy
   * @param {SolanaNetwork} solanaNetwork - Solana network
   * @param {Keypair} keypair - Default keypair to use for paying transaction costs
   */
  private constructor(
    programs: Program[],
    solanaNetwork: SolanaNetwork,
    owner: PublicKey,
    keypair?: Keypair
  ) {
    this._connection = programs[0].provider.connection
    this._programs = programs
    this._provider = programs[0].provider
    this._solanaNetwork = solanaNetwork
    this._quoteTokenMint = getQuoteToken(solanaNetwork)
    this._keypair = keypair
    this._owner = owner
  }

  /**
   * Initialises an AverClient object
   *
   * @param {Connection} connection - Solana Client Object
   * @param {SolanaNetwork} solanaNetwork - Solana network
   * @param {PublicKey | null} owner - Default keypair to pay transaction costs (and rent costs) unless one is otherwise specified for a given transaction.
   * @param {ConfirmOptions} opts - Default options for sending transactions.
   * @param {PublicKey} programIds - Program public key. Defaults to AVER_PROGRAM_IDS.
   * @returns {AverClient} - The Aver Client object
   */
   static async loadAverClient(
    connection: Connection,
    solanaNetwork: SolanaNetwork = SolanaNetwork.Devnet,
    owner: null | PublicKey | Keypair,
    opts?: ConfirmOptions,
    programIds: PublicKey[] = AVER_PROGRAM_IDS
  ) {
    let wallet: NodeWallet;
    let keypair: Keypair
    let pubkey: PublicKey = null

    if (owner instanceof Keypair) {
      // create a wallet with the keypair
      wallet = new NodeWallet(owner);
      keypair = owner
      pubkey = owner.publicKey
    } else if (owner instanceof PublicKey) {
      // create a dummy wallet
      keypair = new Keypair()
      wallet = new NodeWallet(keypair);
      pubkey = owner
    } else {
      // create a dummy wallet
      keypair = new Keypair()
      wallet = new NodeWallet(keypair);
      pubkey = keypair.publicKey
    }

    const provider = new Provider(
      connection,
      wallet,
      opts || {
        commitment: connection.commitment,
        preflightCommitment: connection.commitment,
        skipPreflight: false,
      }
    );

    const programs = await Promise.all(programIds.map(programId => AverClient.loadProgram(provider, programId)))
    return new AverClient(programs, solanaNetwork, pubkey, keypair)
  }

  get connection() {
    return this._connection
  }

  get programs() {
    return this._programs
  }

  /**
   * DEPRECATE
   */
  get program() {
    return this._programs[0]
  }

  get solanaNetwork() {
    return this._solanaNetwork
  }

  get quoteTokenMint() {
    return this._quoteTokenMint
  }

  get keypair() {
    return this._keypair
  }

  get owner() {
    return this._owner
  }

  static async loadProgram(provider: Provider, programId: PublicKey) {
    const idl = await Program.fetchIdl(programId, provider)
    if (idl) {
      const program = new Program(idl, programId, provider)
      checkIdlHasSameInstructionsAsSdk(program)
      return program
    } else {
      throw new Error("Program IDL not loaded")
    }
  }

  async addProgram(programId: PublicKey){
    const program = await AverClient.loadProgram(this._provider, programId)
    this._programs.push(program)
    return program
  }

  async getProgramFromProgramId(programId: PublicKey){
    for (const program of this._programs) {
      if (program.programId.equals(programId)) {
        return program
      }
    }

    return await this.addProgram(programId)
  }

  /**
   * Formats the instruction to create an associated token account
   *
   * @param {PublicKey} mint - Associated token account mint
   * @param {PublicKey} owner - Owner of the assocaited token account
   * @param {PublicKey} payer - Fee payer for creating the associated token account
   *
   * @returns {Promise<TransactionInstruction>} - Instruction to create an associated token account
   */
  async createTokenAtaInstruction(
    mint: PublicKey = this.quoteTokenMint,
    owner: PublicKey = this.owner,
    payer: PublicKey = this.owner
  ) {
    const ataAddress = await getAssociatedTokenAddress(
      mint,
      owner,
      undefined,
      TOKEN_PROGRAM_ID,
      ASSOCIATED_TOKEN_PROGRAM_ID
    )

    return createAssociatedTokenAccountInstruction(
      payer,
      ataAddress,
      owner,
      mint,
      TOKEN_PROGRAM_ID,
      ASSOCIATED_TOKEN_PROGRAM_ID
    )
  }

  /**
   * Attempts to load an Associate Token Account (ATA) or creates one if not found.
   *
   * @param {PublicKey} payer - Fee payer for creating the associated token account
   * @param {PublicKey} mint - Associated token account mint
   * @param {PublicKey} owner - Owner of the assocaited token account
   *
   * @returns {Promise<Account>} - ATA Account
   */
  async getOrCreateTokenAta(
    payer: Signer,
    mint: PublicKey = this.quoteTokenMint,
    owner: PublicKey = this.owner
  ) {
    return getOrCreateAssociatedTokenAccount(
      this._connection,
      payer,
      mint,
      owner
    )
  }

  /**
   * Request an airdrop of lamports (SOL). This method is only supported on devnet.
   *
   * @param {number} amount - Lamports to airdrop. Note 1 lamport = 10^-9 SOL. Max of 1 SOL (10^9 lamports) applies.
   * @param {PublicKey} owner - Public key of account to be airdropped
   *
   * @returns {Promise<string>} - RPC Response Signature
   */
  async requestLamportAirdrop(
    amount: number = LAMPORTS_PER_SOL,
    owner = this.owner
  ) {
    if (this.solanaNetwork == SolanaNetwork.Mainnet) {
      throw new Error("Cannot airdrop on mainnet")
    }

    return this._connection.requestAirdrop(owner, amount)
  }

  /**
   * Fetches the balance for an Associated Token Account (ATA).
   *
   * Note: the value returned is the integer representation of the balance, where a unit is the smallest possible increment of the token.
   * For example, USDC is a 6 decimal place token, so a value of 1,000,000 here = 1 USDC.
   *
   * @param {PublicKey} ata - ATA public key
   *
   * @returns {Promise<RpcResponseAndContext<TokenAmount>>} - Token balance
   */
  async requestAtaBalance(ata: PublicKey) {
    return await this.connection.getTokenAccountBalance(ata)
  }

  /**
   * Fetches a wallet's token balance, given the wallet's owner and the token mint's public key.
   *
   * Note: the value returned is the integer representation of the balance, where a unit is the smallest possible increment of the token.
   * For example, USDC is a 6 decimal place token, so a value of 1,000,000 here = 1 USDC.
   *
   * @param {PublicKey} mint The public key of the token mint
   * @param {PublicKey} owner The public key of the wallet
   *
   * @returns {Promise<RpcResponseAndContext<TokenAmount>>} - Token balance
   */
  async requestTokenBalance(
    mint: PublicKey = this.quoteTokenMint,
    owner: PublicKey = this.owner
  ) {
    const ata = await getAssociatedTokenAddress(mint, owner)
    return (await this.requestAtaBalance(ata))?.value
  }

  /**
   * Fetches Lamport (SOL) balance for a given wallet
   *
   * Note: the value returned is the integer representation of the balance, where a unit is one lamport (=10^-9 SOL)
   * For example, a value of 1,000,000,000 (lamports) = 1 SOL.
   *
   * @param {PublicKey} owner - The wallet owner
   *
   * @returns {Promise<number>} - Lamport balanace
   */
  async requestLamportBalance(owner: PublicKey) {
    return this.connection.getBalance(owner)
  }

  /**
   * Loads current solana system datetime
   *
   * @returns {Promise<number | null>} - Current Solana Clock time
   */
  async getSystemClockDatetime() {
    const slot = await this.connection.getSlot()
    const time = await this.connection.getBlockTime(slot)
    return time ? time * 1000 : null
  }

  /**
   * 
   * @returns DEPCREATED
   */
  async checkHealth() {
    const url = "https://dev.api.aver.exchange" + "/health" + "?format=json"
    let averHealthResponse = await axios
      .get(url)
      .then((res) => res.status.toString().startsWith("2"))
      .catch((_e) => false)
    let solanaHealthResponse = await this.connection
      .getVersion()
      .then((res) => !!res["solana-core"])
      .catch((_e) => false)

    return {
      api: averHealthResponse,
      solana: solanaHealthResponse,
    }
  }

    /**
   * DEPCREATED
   */
  async requestTokenAirdrop(
    amount = 1_000_000_000,
    mint: PublicKey = this.quoteTokenMint,
    owner: PublicKey = this.owner
  ) {
    if (this.solanaNetwork == SolanaNetwork.Mainnet) {
      throw new Error("Cannot airdrop on mainnet")
    }

    const url = "https://dev.api.aver.exchange" + "/airdrop"

    const params = {
      wallet: owner.toBase58(),
      mint: mint,
      amount: amount,
    }

    return axios.post(url, params)
  }

}